numpy/_core/src/multiarray/arraytypes.c.src

```diff
@@ -4318,10 +4320,202 @@ PyArray_DescrFromType(int type)
+ * values (not a true IEEE quad), giving ~106 bits of mantissa precision (53+53).
+ * The correct epsilon is 2^-105, but glibc incorrectly returns 2^-1074 (DBL_TRUE_MIN).
+ */
+#if defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_LE) || \
+    defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_BE)
+    #undef LDBL_EPSILON
+    #define LDBL_EPSILON 0x1p-105L  /* 2^-105 */
+#endif
+
+/*
+ * Define *_TRUE_MIN macros for smallest subnormal values if not available.
+ * Use nextafter(0, 1) to get the smallest positive representable value.
+ */
+#ifndef FLT_TRUE_MIN
+    #define FLT_TRUE_MIN npy_nextafterf(0.0f, 1.0f)
+#endif
+#ifndef DBL_TRUE_MIN
+    #define DBL_TRUE_MIN npy_nextafter(0.0, 1.0)
+#endif
+#ifndef LDBL_TRUE_MIN
+    #define LDBL_TRUE_MIN npy_nextafterl(0.0L, 1.0L)
+#endif
+
+/**begin repeat
+ * #NAME = HALF,FLOAT, DOUBLE, LONGDOUBLE#
+ * #ABB = HALF, FLT, DBL, LDBL#
+ * #type = npy_half, npy_float, npy_double, npy_longdouble#
+ * #RADIX = 16384, 2, 2, 2#
+ * #NEG_MAX = HALF_NEG_MAX, -FLT_MAX, -DBL_MAX, -LDBL_MAX#
+ */
+static int
+@NAME@_get_constant(PyArray_Descr *descr, int constant_id, void *ptr)
+{
+    @type@ val;
+    switch (constant_id) {
+        case NPY_CONSTANT_zero:
+            val = 0;
+            break;
+        case NPY_CONSTANT_one:
+            val = 1;
+            break;
+        case NPY_CONSTANT_minimum_finite:
+            val = @NEG_MAX@;
+            break;
+        case NPY_CONSTANT_maximum_finite:
+            #if defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_LE) || \
+                defined(HAVE_LDOUBLE_IBM_DOUBLE_DOUBLE_BE)
+                /* For IBM double-double, use nextafter(inf, 0) to get the true
+                 * maximum representable value (matches old MachArLike behavior) */
+                if (sizeof(@type@) == sizeof(npy_longdouble)) {
+                    val = npy_nextafterl((@type@)NPY_INFINITY, (@type@)0.0L);
+                    break;
+                }
+            #endif
+            val = @ABB@_MAX;
+            break;
+        case NPY_CONSTANT_inf:
+            val = (@type@)NPY_INFINITYF;
+            break;
+        case NPY_CONSTANT_nan:
+            val = (@type@)NPY_NANF;
+            break;
+        case NPY_CONSTANT_finfo_radix:
+            val = @RADIX@;
+            break;
+        case NPY_CONSTANT_finfo_eps:
+            val = @ABB@_EPSILON;
+            break;
+        case NPY_CONSTANT_finfo_smallest_normal:
+            val = @ABB@_MIN;
+            break;
+        case NPY_CONSTANT_finfo_smallest_subnormal:
+            val = @ABB@_TRUE_MIN;
+            break;
+        case NPY_CONSTANT_finfo_nmant:
+            *(npy_intp *)ptr = @ABB@_MANT_DIG - 1;
+            return 1;
+        case NPY_CONSTANT_finfo_min_exp:
+            /*
+            Definition: Minimum negative integer such that FLT_RADIX raised by power one less than that integer is a normalized float, double and long double respectively
+
+            refernce: https://en.cppreference.com/w/c/types/limits.html
```

Comment:
I left these comments for developers so that they can reference why we are subtracting 1